WalletWasabi/Crypto/Api/CredentialIssuer.cs

```diff
@@ -0,0 +1,146 @@
+			}
+
+			var statements = new List<Statement>();
+			foreach (var presentation in registrationRequest.Presented)
+			{
+				// Calculate Z using coordinator secret
+				var Z = presentation.ComputeZ(CoordinatorSecretKey);
+
+				statements.Add(ProofSystem.ShowCredential(presentation, Z, CoordinatorParameters));
+
+				// Check if the serial numbers have been used before. Note that
+				// the serial numbers have not yet been verified at this point, but a
+				// request with an invalid proof and a used serial number should also be
+				// rejected.
+				if (SerialNumbers.Contains(presentation.S))
+				{
+					throw new WabiSabiException(WabiSabiErrorCode.SerialNumberAlreadyUsed, $"Serial number reused {presentation.S}");
+				}
+			}
+
+			foreach (var credentialRequest in registrationRequest.Requested)
+			{
+				statements.Add(registrationRequest.IsNullRequest
+					? ProofSystem.ZeroProof(credentialRequest.Ma)
+					: ProofSystem.RangeProof(credentialRequest.Ma, credentialRequest.BitCommitments));
+			}
+
+			// Balance proof
+			if (!registrationRequest.IsNullRequest)
+			{
+				var presented = registrationRequest.Presented.Select(x => x.Ca).Sum();
+				var requested = registrationRequest.Requested.Select(x => x.Ma).Sum();
+
+				// A positive Delta_a means the requested credential amounts are larger
+				// than the presented ones (i.e. input registration, and a negative
+				// balance corresponds to output registration). The equation requires a
+				// commitment to 0, so the sum of the presented attributes and the
+				// negated requested attributes is tweaked by delta_a.
+				var absAmountDelta = new Scalar(registrationRequest.Balance.Abs());
+				var deltaA = registrationRequest.Balance < Money.Zero ? absAmountDelta.Negate() : absAmountDelta;
+				var balanceTweak = deltaA * Generators.Gg;
+				statements.Add(ProofSystem.BalanceProof(balanceTweak + presented - requested));
+			}
+
+			var transcript = new Transcript(new byte[0]); // FIXME label unified registration, K, isNullRequest
+
+			// Verify all statements.
+			var areProofsValid = Verifier.Verify(transcript, statements, registrationRequest.Proofs);
+			if (!areProofsValid)
+			{
+				throw new WabiSabiException(WabiSabiErrorCode.CoordinatorReceivedInvalidProofs);
+			}
+
+			// Issue credentials.
+			var credentials = registrationRequest.Requested.Select(x => IssueCredential(x.Ma, Random.GetScalar())).ToArray();
+
+			// Construct response.
+			var responseTranscript = new Transcript(new byte[0]); // FIXME label unified registration, K, isNullRequest
+
+			var proofs = Prover.Prove(responseTranscript, credentials.Select(x => x.Knowledge), Random);
+			var macs = credentials.Select(x => x.Mac);
+			var response = new RegistrationResponse(macs, proofs);
+
+			// Register the serial numbers to prevent credential reuse.
+			foreach (var presentation in registrationRequest.Presented)
+			{
+				SerialNumbers.Add(presentation.S);
+			}
+			Balance += registrationRequest.Balance;
```

Comment:
Should not this be` Balance += deltaA;` I guess the `Balance` variable holds the amount of outstanding registered satoshis. Before users registers their output, this variable could have been inflated to an arbitrarily large value just by reissuing credentials. So, I guess here we just want to add `deltaA` to `Balance` to get meaningful values, If I understand correctly.

Reply:
Yes, you're right.